tensorflow_probability/python/experimental/mcmc/particle_filter.py

```diff
@@ -273,19 +273,126 @@ def observation_fn(_, state):
+      If `None`, only values from the final step are returned.
+      Default value: `lambda *_: True` (trace every step).
+    static_trace_allocation_size: Optional Python `int` size of trace to
+      allocate statically. This should be an upper bound on the number of steps
+      traced and is used only when the length cannot be
+      statically inferred (for example, if a `trace_criterion_fn` is
+      specified).
+      It is primarily intended for contexts where static shapes are required,
+      such as in XLA-compiled code.
+      Default value: `None`.
+    parallel_iterations: Passed to the internal `tf.while_loop`.
+      Default value: `1`.
+    seed: PRNG seed; see `tfp.random.sanitize_seed` for details.
+      name: Python `str` name for ops created by this method.
+      Default value: `None` (i.e., `'particle_filter'`).
+  Returns:
+    traced_results: A structure of Tensors as returned by `trace_fn`. If
+      `trace_criterion_fn==None`, this is computed from the final step;
+      otherwise, each Tensor will have initial dimension `num_steps_traced`
+      and stacks the traced results across all steps.
+
+  #### References
+
+  [1] Adam Scibior, Vaden Masrani, and Frank Wood. Differentiable Particle
+      Filtering without Modifying the Forward Pass. _arXiv preprint
+      arXiv:2106.10314_, 2021. https://arxiv.org/abs/2106.10314
+  """
+  kernel = smc_kernel.SequentialMonteCarlo(
+      propose_and_update_log_weights_fn=propose_and_update_log_weights_fn,
+      resample_fn=resample_fn,
+      resample_criterion_fn=resample_criterion_fn,
+      particles_dim=particles_dim,
+      unbiased_gradients=unbiased_gradients
+  )
+
+  # Use `trace_scan` rather than `sample_chain` directly because the latter
+  # would force us to trace the state history (with or without thinning),
+  # which is not always appropriate.
+  def seeded_one_step(seed_state_results, _):
+
+    seed, state, results = seed_state_results
+
+    one_step_seed, next_seed = samplers.split_seed(seed)
+
+    next_state, next_results = kernel.one_step(
+        state, results, seed=one_step_seed)
+
+    return next_seed, next_state, next_results
+
+  final_seed_state_result, traced_results = loop_util.trace_scan(
+      loop_fn=seeded_one_step,
+      initial_state=(loop_seed,
+                     initial_weighted_particles,
+                     kernel.bootstrap_results(initial_weighted_particles)),
+      elems=tf.ones([num_timesteps]),
+      trace_fn=lambda seed_state_results: trace_fn(*seed_state_results[1:]),
+      trace_criterion_fn=(
+          lambda seed_state_results: trace_criterion_fn(  # pylint: disable=g-long-lambda
+              *seed_state_results[1:])),
+      static_trace_allocation_size=static_trace_allocation_size,
+      parallel_iterations=parallel_iterations)
+
+  if trace_criterion_fn is never_trace:
+    # Return results from just the final step.
+    traced_results = trace_fn(*final_seed_state_result[1:])
+
+    return traced_results
```

Comment:
This line needs to be un-indented -- so we return `trace_results` whether or not `trace_criterion_fn is never_trace`.